readme.md
# Yaxy

[Более полное описание на русском языке](readme_ru.md)

Yaxy is a proxy-server for a web developer, it substitutes the required resources with simple rules.

## Installation

If you haven't installed NodeJS already, [then do it] (http://nodejs.org/), then

     npm install -g yaxy

## Running

     yaxy --config my-yaxy-config.txt --port 9999 --proxy user:password@proxyhost:3000

If you do not specify `--config`, Yaxy will look for yaxy-config.txt file in the current directory. The server will listen on port `8558`, if no `--port` specified. If you specify `--proxy`, Yaxy will route external requests through proxy server.

## Configuration File Format

The configuration file is read line by line. Blank lines, lines beginning with `#` and unknown strings are ignored.

Rules can be combined in sections. Start of the section is a line enclosed in square brackets, the contents of a line can be any.

    # Rules out of sections

    [Section 1]
    # Rules for section #1

    [Section 2]
    # Rules for section #2

If a section name starts with `#`, then all rules of this section will be ignored.

    # Rules out of sections

    [#Section 1]
    # Section #1 rules
    # will be ignored

    [Section 2]
    # Rules for section #2

## Rules

Rules are written like:

    url => replacement

`url` -- domain address string for rule to work. `http://` at the beginning of line is not obligaroty. E.g., rule `www.yandex.ru => ...` will work for all resources at www.yandex.ru domain, while `yandex.ru/yandsearch => ...` -- only for Yandex search.

Sometimes exact address match is required, i.e. you need to modify only site's main page, leaving the rest of it intact. For this specify `!` at the beginning of url, `http://` is still optional, but trailing slash(es) are a must for exact matching. E.g., replaceing Yandex main page:  `!www.yandex.ru/ => ...`.


If first two cases are not enough, you may use regexp for left side, enclosed with `/`. So, if you'll need to modify all requests to .ru domain, use:  `/^http://[^/]+\.ru// => ...`. Please note, regexp uses whole url, including `http://`.

### Simple url substitute

The right side of a rule specifies the replacement for captured match of a rule's left side. E.g., rule `google.ru => yandex.ru` will substitute all request from `google.ru` to `yandex.ru`, so request `http://google.ru/` goes to `http://yandex.ru/`, and request to `http://google.ru/foo/bar` goes to `http://yandex.ru/foo/bar`.
More specific rule `google.ru/foo => yandex.ru/foo` will substitute request to `http://google.ru/foo/bar`, while leaving intact `http://google.ru/baz`. Please note, if you need to keep path in a domain, you must specify it on both parts of a rule, because `google.ru/foo => yandex.ru` will replace request `http://google.ru/foo/bar` to `http://yandex.ru/bar`.

When you have a regexp in a left side of a rule, right side is treated as a template for captured groups. E.g., we replace all *.ru domain requests to *.com: `/^(http://[^/]+\.)ru(/.*)$/ => $1com$2`.

### Displaying local files

You can have your little static-server, using `file://` protocol in the right part of a rule.

    # windows
    host.com/some/path => file://c:/www/host.com
    # linux
    host.com/some/path => file:///home/www/host.com

Directory index file -- index.html. For above example, url `http://host.com/some/path` will return file `/home/www/host.com/index.html`, and url `http://host.com/some/path/empty.gif` returns `/home/www/host.com/empty.gif`.

### Using data:uri

If your substitute is simple and you don't want to create a file, you may use data:uri.

    host.com/some/path => data:text/html;<script type="text/javascript">alert('Hello!');</script>

Then all urls, starting with `host.com/some/path` will return response with `<script type="text/javascript">alert('Hello!');</script>`.

    host.com/some/path/png => data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAABAAAAAQAQMAAAAlPW0iAAAABlBMVEUAAAD///+l2Z/dAAAAM0lEQVR4nGP4/5/h/1+G/58ZDrAz3D/McH8yw83NDDeNGe4Ug9C9zwz3gVLMDA/A6P9/AFGGFyjOXZtQAAAAAElFTkSuQmCC

Get an image for all urls, starting with `host.com/some/path/png`

If you have a regexp in rule's left side, you may use it's captured groups in the right side.

    /^http://test.my/\?name=(.*)/ => data:text/html;<script>alert('Hello, $1!');</script>

### Routing requests through proxy

Beside the specifying `--proxy` on the start, Yaxy can route specific requests through other proxy.

    www.yandex.ru => proxy:localhost:3000

With this rule requests to Yandex will be routed through localhost proxy, that listens port number 3000. If Yaxy was started with `--proxy localhost:3333` arg, all requests will be routed through localhost default proxy (port 3333), but requests to Yandex will still use route through proxy on port number 3000.

### Leaving url intact

If url must be left intact, use `$` in rule's right side.

    host.com/some/path => $

Sometimes you don't need to modify a request, but only use some modifiers for it (see below).

### Cancelling a request

If right side of rule is left empty, then request will be cancelled (even no http headers will be returned).

## Modifiers

Modifiers are strings, starting with `$` (leading spaces don't matter). First goes modifier name, after space -- arguments. Modifier corresponds to the rule it is written after. If modifier is written at the file's beginning (before any rule), it applies to all rules. If modifier is written at the section's beginning, then it applies to all section rules.

Please note, global modifiers applies to *rules* (i.e. requests, matching these rules), and not all requests through proxy server.

### Modifying GET-parameters

    # Add or modify parameter
    $SetQueryParam from=yaxy

    # Removing parameter
    $RemoveQueryParam from

### Cookie manipulations

Cookies are modified for server, so while browser sends an original cookie, server receives the modified one.

    # Adding/modifying specific cookie
    $SetCookie user=YaxyUser

    # Removing a cookie
    $RemoveCookie ssid

### Modifying HTTP headers

    # Setting a request header
    $SetRequestHeader X-Requested-With: Yaxy

    # Setting a response header
    $SetResponseHeader X-Proxy: Yaxy

    # Removing request header
    $RemoveRequestHeader Referer

    # Removing response header
    $RemoveResponseHeader Content-Type

### Setting root directory for `file://` protocol

If all of your projects are located in one directory, you can set root directory one time, instead of doing it each time for every project.

    $SetDocumentRoot /home/me/projects

Now you can use this path with tilde (`~`)

    site.my => file://~/site.my
    # With the previous rule it is equal to
    site => file:///home/me/projects/site.my

Also you can set root directory for specific section of rules

    $SetDocumentRoot /home/me/projects

    [site.my]
    $SetDocumentRoot ~/site.my
    site.my/css => file://~/css-dev
    site.my/js => file://~/js-dev
    site.my => file://~    

### Modifying response status

    # Always HTTP 200 for response (even when it's not)
    $StatusCode 200

### Response delay

    #Respond at least 5 seconds later
    $Delay 5


changelog.md
### Version 0.3.0

* WebSocket support

### Version 0.2.0

* $Include

### Version 0.1.15

* Custom SSL sertificates

### Version 0.1.14

* Headers bugfix

### Version 0.1.12

* state.removeCookie bugfix

### Version 0.1.11

* API improvements

### Version 0.1.10

* rejectUnauthorized https option

### Version 0.1.9

* Bugfixes

### Version 0.1.8

* Full https support

### Version 0.1.7

* Simple https support

### Version 0.1.6

* Punycode support

### Version 0.1.5

* Templates
* `bin:` rules

### Version 0.1.4

* Proxy support


test/index.js
var spawn = require('child_process').spawn;
var assert = require('assert');

describe('Yaxy', function() {
    require('../tools/test-server');

    describe('Rules', function() {
        var tests = [
            ['simple-request', 'http://localhost:9595/', 'Hello'],
            ['simple-rewrite', 'http://www.yandex.ru/', 'Hello'],
            ['include-01', 'http://www.yandex.ru/', 'Hello'],
            ['include-02', 'http://www.yandex.ru/', 'Hello'],
            ['include-03', 'http://www.yandex.ru/', 'Hello']
        ];

        var yaxy;

        afterEach(function() {
            yaxy.kill();
        });
        this.timeout(5000);

        tests.forEach(function(test) {
            it(test[0], function() {
                return new Promise(function(resolve, reject) {
                    try {
                        yaxy = spawn('node', ['index.js', '--config', 'test/yaxy-configs/' + test[0] + '.txt', '--port', '9559', '--https-port', '9569'], {
                            cwd: process.cwd()
                        });
                        yaxy.stdout.on('data', function(chunk) {
                            console.log('Yaxy: ' + chunk);
                        });
                        yaxy.stderr.on('data', function(chunk) {
                            yaxy.kill();
                            reject('YaxyError: ' + chunk);
                        });
                    } catch (err) {
                        reject(err);
                    }
                    setTimeout(function() {
                        var phantomResult = '';
                        var phantomError = '';
                        var phantom = spawn(require('phantomjs').path, ['--proxy=localhost:9559', 'tools/phantom.js', test[1]]);
                        phantom.stdout.on('data', function(chunk) {
                            phantomResult += chunk;
                        });
                        phantom.stderr.on('data', function(chunk) {
                            phantomError += chunk;
                        });
                        phantom.on('exit', function() {
                            if (phantomError) {
                                reject(new Error('PhantomError: ' + phantomError));
                            } else {
                                assert.equal(phantomResult.trim(), test[2]);
                                resolve();
                            }
                        });
                    }, 500);
                });
            });
        });
    });
});


tests/cli-tests.js
function trim(text) {
    return (text || "").replace(/^\s+|\s+$/g, "");
}

test('trim()', function () {
    equal(trim(''), '', 'Пустая строка');
    ok(trim('   ') === '', 'Строка из пробельных символов');
    deepEqual(trim(), '', 'Без параметра');

    equal(trim(' x'), 'x', 'Начальные пробелы');
    equal(trim('x '), 'x', 'Концевые пробелы');
    equal(trim(' x '), 'x', 'Пробелы с обоих концов');
    equal(trim('    x  '), 'x', 'Табы');
    equal(trim('    x   y  '), 'x   y', 'Табы и пробелы внутри строки не трогаем');
});

lib/yaxy.js
var fileExists = require('fs').exists || require('path').exists;

module.exports = Yaxy;

function Yaxy(port, httpsPort, certs) {
    if (!(this instanceof Yaxy)) {
        return new Yaxy(port, httpsPort, certs);
    }

    certs = certs || {};

    this._listeners = [];
    this._proxy = null;
    this._auth = null;
    this._sslHosts = [];
    this._httpsPort = httpsPort;
    this._httpPort = port;

    this._server = require('http').createServer(this._onRequest.bind(this, 'http:'));
    this._server.on('connect', this._onConnect.bind(this));
    this._server.on('upgrade', this._onRequest.bind(this, 'ws:'));
    this._server.listen(port);

    this._httpsServer = null;
    if (httpsPort) {
        var read = require('fs').readFileSync;
        var resolve = require('path').resolve;
        var options = {
            key: read(resolve(certs.key || (__dirname + '/../ssl/privatekey.pem'))),
            cert: read(resolve(certs.cert || (__dirname + '/../ssl/certificate.pem')))
        };
        this._httpsServer = require('https').createServer(options, this._onRequest.bind(this, 'https:'));
        this._httpsServer.on('upgrade', this._onRequest.bind(this, 'wss:'));
        this._httpsServer.listen(httpsPort);
    }
}

Yaxy.prototype = {
    /**
     *
     * @param {Object} listener Объект с возможными полями
     *      * method
     *      * url
     *      * urlStart
     *      * host
     *      * rhost
     *      * path
     *      * pathStart
     *      * queryParams {Object}
     *      * headers {Object}
     *      * referer
     *      * refererStart
     *      * refererHost
     *      * refererRhost
     *      * refererPath
     *      * refererPathStart
     *      * refererQueryParams
     *      * cookies {Object}
     *      * fn
     */
    bind: function(listener) {
        this._listeners.push(listener);
    },

    unbind: function(listener) {
        var index = this._listeners.indexOf(listener);
        if (index > -1) {
            this._listeners.splice(index, 1);
        }
    },

    unbindAll: function() {
        this._listeners.length = 0;
    },

    useSSLFor: function(host) {
        if (host.indexOf(':') === -1) {
            host += ':443';
        }
        if (this._sslHosts.indexOf(host) === -1) {
            this._sslHosts.push(host);
        }
    },

    unuseSSLFor: function(host) {
        if (host.indexOf(':') === -1) {
            host += ':443';
        }
        var index = this._sslHosts.indexOf(host);
        if (index > -1) {
            this._sslHosts.splice(index, 1);
        }
    },

    unuseAllSSL: function() {
        this._sslHosts.length = 0;
    },

    setProxy: function(proxy) {
        this._proxy = proxy;
    },

    setAuth: function(userPasswd) {
        this._auth = 'Basic ' + new Buffer(userPasswd).toString('base64');
    },


    _onRequest: function(protocol, req, res) {
        if (this._auth && req.headers['proxy-authorization'] !== this._auth) {
            res.writeHead(407, 'Proxy Authentication Required', {
                'Proxy-Authenticate': 'Basic'
            });
            res.end();
            return;
        }

        var state = new State(protocol, req, res);
        state.setProxy(this._proxy);
        (function next(i) {
            if (i < this._listeners.length) {
                state._match(this._listeners[i], function() {
                    next.call(this, i + 1);
                }, this);
            } else {
                state.doRequest();
            }
        }).call(this, 0);
    },

    _onConnect: function(request, socketRequest, head) {
        if (this._auth && request.headers['proxy-authorization'] !== this._auth) {
            socketRequest.write("HTTP/" + request.httpVersion + " 407 Proxy Authentication Required\r\nProxy-Authenticate: Basic\r\n\r\n");
            socketRequest.end();
            return;
        }

        var ph = require('url').parse('http://' + request.url);
        var host = 'localhost';
        var port = this._httpsPort;
        if (ph.port === '80') {
            port = this._httpPort;
        } else if (!port || this._sslHosts.indexOf(request.url) === -1) {
            host = ph.hostname;
            port = Number(ph.port);
        }
        var socket = require('net').connect(port, host, function() {
            socket.write(head);
            socketRequest.write("HTTP/" + request.httpVersion + " 200 Connection established\r\n\r\n");
        });

        socket.on('data', function(chunk) {
            socketRequest.write(chunk);
        });
        socket.on('end', function() {
            socketRequest.end();
        });
        socket.on('error', function() {
            socketRequest.write("HTTP/" + request.httpVersion + " 500 Connection error\r\n\r\n");
            socketRequest.end();
        });

        socketRequest.on('data', function(chunk) {
            socket.write(chunk);
        });
        socketRequest.on('end', function() {
            socket.end();
        });
        socketRequest.on('error', function() {
            socket.end();
        });
    }
};

function State(protocol, req, res, head) {
    this._req = req;
    this._res = res;
    this._requestHead = head;
    this._proxy = null;

    this._requestBodyChunks = [];
    this._isRequestEnd = false;
    this._requestEndListeners = [];
    this._req.on('data', this._onRequestData.bind(this));
    this._req.on('end', this._onRequestEnd.bind(this));

    this._responseStatus = 0;
    this._requestHeaders = this._req.headers;
    this._responseHeadersOverwrites = {};

    if (!/^(http|ws)s?:\/\//.test(this._req.url)) {
        this._req.url = protocol + '//' + this._req.headers['host'] + this._req.url;
    }
    this.setRequestUrl(this._req.url);
    this._overwriteQueryParams = {};
    this._referer = this._req.headers['referer'] ? require('url').parse(this._req.headers['referer']) : {
        host: '',
        pathname: '',
        search: ''
    };
    this._refererQueryParams = this._referer.search ? require('querystring').parse(this._referer.search) : {};
    this._refererRhost = this._referer.host ? this._referer.host.split('.').reverse() : [];

    this._cookies = this._parseCookie(this._req.headers['cookie']);

    this._vars = {};
}

State.prototype = {
    getMethod: function() {
        return this._req.method;
    },

    getRequestUrl: function() {
        return this._urlString;
    },

    getRequestProtocol: function() {
        return this._url.protocol;
    },

    getRequestHost: function() {
        return this._url.hostname;
    },

    getRequestPort: function() {
        return this._url.port;
    },

    getRequestPath: function() {
        return this._url.pathname;
    },

    getQueryString: function() {
        return this._url.query;
    },

    setRequestUrl: function(url) {
        this._urlString = url;
        this._url = require('url').parse(url);
        this._queryParams = require('querystring').parse(this._url.query);
        this._rhost = this._url.host.split('.').reverse();
    },

    overwriteResponseStatus: function(status) {
        this._responseStatus = status;
    },

    setRequestHeader: function(header, value) {
        this._requestHeaders[header.toLowerCase()] = String(value);
    },

    getRequestHeader: function(header) {
        return this._requestHeaders[header.toLowerCase()] || '';
    },

    removeRequestHeader: function(header) {
        delete this._requestHeaders[header.toLowerCase()];
    },

    setResponseHeader: function(header, value) {
        this._responseHeadersOverwrites[header.toLowerCase()] = String(value);
    },

    removeResponseHeader: function(header) {
        this._responseHeadersOverwrites[header.toLowerCase()] = undefined;
    },

    setResponseType: function(contentType) {
        if (!this._responseHeadersOverwrites['content-type']) {
            if (contentType.substr(0, 4) === 'text' || contentType.indexOf('xml') >= 0 || contentType.indexOf('json') >= 0 || contentType.indexOf('javascript') >= 0) {
                contentType += '; charset=UTF-8';
            }
            this.setResponseHeader('Content-Type', contentType);
        }
    },

    setQueryParam: function(param, value) {
        this._overwriteQueryParams[param] = value;
    },

    getQueryParam: function(param) {
        return require('querystring').parse(this._url.query)[param] || '';
    },

    removeQueryParam: function(param) {
        this._overwriteQueryParams[param] = null;
    },

    setCookie: function(name, value) {
        this._cookies[name] = value;
    },

    getCookie: function(name) {
        return this._cookies[name];
    },

    removeCookie: function(name) {
        delete this._cookies[name];
    },

    getRequestBody: function(encoding) {
        return Buffer.concat(this._requestBodyChunks).toString(encoding || 'utf8');
    },

    redirect: function(url, status) {
        this.overwriteResponseStatus(status || 302);
        this.setResponseHeader('Location', url);
        this.send('');
    },

    doRequest: function() {
        var protocol = this._url.protocol || 'http:';
        if (!/^(http|ws)s?:$/.test(protocol)) {
            return this.error(new Error('Unknown protocol ' + protocol));
        }
        var secure = protocol === 'https:' || protocol === 'wss:';
        var host = this._url.hostname || this._requestHeaders['host'];
        var port = +(this._url.port || host.split(':')[1] || (secure ? 443 : 80));
        var queryString = this._buildQueryString();
        var path = this._url.pathname + (queryString ? '?' + queryString : '');
        var cookieString = this._buildCookie(this._cookies);
        if (cookieString) {
            this.setRequestHeader('Cookie', cookieString);
        } else {
            this.removeRequestHeader('Cookie');
        }
        if (this._requestHeaders.connection === 'keep-alive') {
            delete this._requestHeaders.connection;
        }
        if (this._proxy && this._proxy.user) {
            this._requestHeaders['proxy-authorization'] = 'Basic ' + new Buffer(this._proxy.user + ':' + this._proxy.password).toString('base64');
        } else {
            delete this._requestHeaders['proxy-authorization'];
        }

        var headers = {};
        Object.keys(this._requestHeaders).forEach(function(name) {
            var camelName = name.replace(/(^|-)([a-z])/g, function(ignored, dash, letter) {
                return dash + letter.toUpperCase();
            });
            headers[camelName] = this._requestHeaders[name];
        }, this);

        var hostByProxy = this._proxy ? this._proxy.host : host;
        var portByProxy = this._proxy ? this._proxy.port : port;
        if (protocol === 'http:' || protocol === 'https:') {
            var requestOptions = {
                host: hostByProxy,
                port: portByProxy,
                method: this._req.method,
                path: this._proxy ? protocol + '//' + host + path : path,
                headers: headers,
                rejectUnauthorized: false
            };

            var clientReq = require(secure ? 'https' : 'http').request(requestOptions, (clientRes) => {
                var statusCode = this._responseStatus || clientRes.statusCode;
                var headers = this._buildResponseHeaders(clientRes.headers);
                this._res.writeHead(statusCode, headers);
                clientRes.pipe(this._res);
            });

            clientReq.on('error', (err) => {
                this.error(err);
            });

            this.ready(function() {
                this._requestBodyChunks.forEach((chunk) => {
                    clientReq.write(chunk);
                });
                clientReq.end();
            }, this);

            if (!this._req.connection.__yaxy_clientReqs) {
                this._req.connection.__yaxy_clientReqs = [];
                this._req.connection.once('close', () => {
                    this._req.connection.__yaxy_clientReqs.forEach(function(req) {
                        req.abort();
                    });
                });
            }
            this._req.connection.__yaxy_clientReqs.push(clientReq);
        } else if (protocol === 'ws:' || protocol === 'wss:') {
            var websocket = require('websocket-driver');
            if (websocket.isWebSocket(this._req)) {
                var toclient = websocket.http(this._req);
                toclient.io.write(this._requestHead);
                this._res.pipe(toclient.io).pipe(this._res);

                var toserver = websocket.client(this._url);
                Object.keys(headers).forEach((header) => {
                    toserver.setHeader(header, headers[header]);
                });
                toclient.messages.pipe(toserver.messages).pipe(toclient.messages);

                var tcp = require('net').connect(portByProxy, hostByProxy, () => {
                    tcp.pipe(toserver.io).pipe(tcp);
                    toclient.start();
                    toserver.start();
                });
            }
        }
    },

    sendFile: function(fname) {
        var that = this;
        fileExists(fname, function(exists) {
            if (exists) {
                require('fs').stat(fname, function(err, stat) {
                    if (err) {
                        return that.error(err);
                    }
                    if (stat.isDirectory()) {
                        that.sendFile(require('path').join(fname, 'index.html'));
                    } else if (stat.isFile()) {
                        that.setResponseType(require('./mime').getType(require('path').extname(fname)));
                        require('fs').readFile(fname, function(err, content) {
                            if (err) {
                                return that.error(err);
                            }
                            that.send(content);
                        });
                    } else {
                        that.abort();
                    }
                });
            } else {
                that.overwriteResponseStatus(404);
                that.setResponseHeader('Content-Type', 'text/plain; charset=UTF-8');
                that.send('Yaxy did not find file ' + fname, 'utf8');
            }
        });
    },

    send: function(content, encoding) {
        this._res.writeHead(this._responseStatus || 200, this._buildResponseHeaders({}));
        this._res.write(content, encoding || 'binary');
        this._res.end();
    },

    abort: function() {
        this._res.end();
    },

    next: function() {
        if (this._matchCallback) {
            this._matchCallback.call(this._matchCtx);
            this._matchCallback = null;
        }
    },

    ready: function(fn, ctx) {
        if (this._isRequestEnd) {
            fn.call(ctx);
        } else {
            this._requestEndListeners.push([fn, ctx]);
        }
    },

    setProxy: function(proxy) {
        this._proxy = proxy;
    },

    set: function(name, value) {
        this._vars[name] = value;
    },

    get: function(name, defaultValue) {
        return name in this._vars ? this._vars[name] : defaultValue;
    },

    error: function(err) {
        if (err.code === 'ECONNRESET') {
            return;
        }
        this.overwriteResponseStatus(500);
        this.setResponseHeader('X-Yaxy-Error', err.message);
        this.send('');
        if (err.code === 'ENOTFOUND') {
            console.log('Host ' + this._url.hostname + ' not found.');
        } else {
            console.error(err.stack);
        }
    },


    _buildResponseHeaders: function(origin) {
        var over = this._responseHeadersOverwrites;
        Object.keys(over).forEach(function(header) {
            if (typeof over[header] == 'undefined') {
                delete origin[header];
            } else {
                origin[header] = over[header];
            }
        });
        return origin;
    },

    _buildQueryString: function() {
        var queryString = this._url.query || '';
        Object.keys(this._overwriteQueryParams).forEach(function(param) {
            var value = this._overwriteQueryParams[param];
            var pattern = new RegExp('(^|&)' + param + '\\b[^&]*');
            queryString = queryString.replace(pattern, value == null ? '' : '$1' + param + '=' + value);
            if (value != null && !pattern.test(queryString)) {
                // значит параметра в строке не было, ничего не заменилось, добавляем
                queryString += (queryString ? '&' : '') + param + '=' + encodeURIComponent(value);
            }
        }, this);
        return queryString;
    },

    _parseCookie: function(cookieString) {
        var cookies = {};
        if (cookieString) {
            cookieString.split(';').forEach(function(oneCookieString) {
                var oneCookiePair = oneCookieString.split('=');
                cookies[oneCookiePair.shift().trim()] = oneCookiePair.join('=').trim();
            }, this);
        }
        return cookies;
    },

    _buildCookie: function(cookies) {
        return Object.keys(cookies).map(function(name) {
            return name + '=' + cookies[name];
        }).join('; ');
    },

    _onRequestData: function(chunk) {
        this._requestBodyChunks.push(chunk);
    },

    _onRequestEnd: function() {
        this._isRequestEnd = true;
        this._req.removeAllListeners();
        this._requestEndListeners.forEach(function(listener) {
            listener[0].call(listener[1]);
        });
    },

    _match: function(listener, callback, ctx) {
        if (!listener.fn) {
            return callback.call(ctx);
        }
        this._matchCallback = callback;
        this._matchCtx = ctx;
        var validListener = Object.keys(listener).every(function(matcherName) {
            if (matcherName.indexOf('_') === 0 || !this._matchers[matcherName]) {
                return true;
            }
            return this._matchers[matcherName].call(this, listener[matcherName]);
        }, this);
        if (validListener) {
            listener.fn.call(listener.ctx, this);
        } else {
            this.next();
        }
    },

    _matchers: {
        method: function(method) {
            return method ? this._matchers._string(method, this._req.method) : true;
        },

        url: function(url) {
            return this._matchers._string(url, this._req.url);
        },

        urlStart: function(url) {
            return this._matchers._stringStart(url, this._req.url);
        },

        host: function(host) {
            return this._url.host === host;
        },

        rhost: function(rhost) {
            return this._matchers._rhost(rhost, this._rhost);
        },

        path: function(path) {
            return this._matchers._string(path, this._url.pathname);
        },

        pathStart: function(path) {
            return this._matchers._stringStart(path, this._url.pathname);
        },

        queryParams: function(params) {
            return this._matchers._object(params, this._queryParams);
        },

        headers: function(headers) {
            return this._matchers._object(headers, this._req.headers);
        },

        referer: function(referer) {
            return this._matchers._string(referer, this._req.headers['referer'] || '');
        },

        refererStart: function(referer) {
            return this._matchers._stringStart(referer, this._req.headers['referer'] || '');
        },

        refererHost: function(host) {
            return this._referer.host === host;
        },

        refererRhost: function(rhost) {
            return this._matchers._rhost(rhost, this._refererRhost);
        },

        refererPath: function(path) {
            return this._matchers._string(path, this._referer.pathname);
        },

        refererPathStart: function(path) {
            return this._matchers._stringStart(path, this._referer.pathname);
        },

        refererQueryParams: function(params) {
            return this._matchers._object(params, this._refererQueryParams);
        },

        cookies: function(cookies) {
            return this._matchers._object(cookies, this._cookies);
        },


        _string: function(pattern, value) {
            return Object.prototype.toString.call(pattern) === '[object RegExp]' ? pattern.test(value) : pattern === value;
        },

        _stringStart: function(pattern, value) {
            return value.indexOf(pattern) === 0;
        },

        _object: function(pattern, value) {
            return Object.keys(pattern).every(function(field) {
                var fieldValue = pattern[field];
                return typeof fieldValue == 'boolean' ? field in value === fieldValue : this._string(fieldValue, value[field] || '');
            }, this);
        },

        _rhost: function(pattern, value) {
            pattern = pattern.replace(/^\.|\.$/g, '').split('.');
            for (var i = 0; i < pattern.length; i++) {
                if (pattern[i] !== value[i]) {
                    return false;
                }
            }
            return true;
        }
    }
};
